bip-0088.mediawiki

```diff
@@ -0,0 +1,229 @@
+# To avoid ambiguity, if a hardened marker appears within any section in the path template, all preceding sections MUST also use the same hardened marker (either "h" or "'").
+# To avoid ambiguity, trailing slashes (for example, "1/2/") and duplicate slashes (for example, "0//1") MUST NOT appear in the template.
+
+It may be desireable to have fully unambiguous encoding, where for each valid path template string, there is no other valid template string that matches the exact same set of paths. This would enable someone to compare templates for equality through a simple string equality check, without any parsing.
+
+To achieve this, two extra rules are needed:
+
+* Within Ranged index template, subsequent range MUST NOT start with the value that is equal to the end of the previous range plus one. Thus, <nowiki>"{1,2,3-5}"</nowiki> is not allowed, and should be specified as <nowiki>"{1-5}"</nowiki> instead. This rule might make templates less convenient for frequent edits, though.
+
+* Only one type of hardened marker should be allowed (either "h" or "'").
+
+Instead of requiring the second extra rule, implementations can simply replace one type of marker with another in the template strings before comparing them.
+
+==Full and partial templates==
+
+If the template starts with "m/", that means that this is the "full" template, that matches the whole path.
+
+If the template does not start with "m/", that means that this is a "partial" template, and it can be used to match a part of the path, in the contexts where this might be appropriate (for example, when constraints for the suffix of the path might be dynamic, while constraints for the prefix of the path are fixed).
+
+Full template can be combined with partial template, where partial template extends full template,
+resulting in new, longer full template.
+
+Partial template can be combined with another partial template, resulting in new, longer partial template.
+
+Full template can not be combined with another full template.
+
+Implementations MUST support parsing full templates and matching paths against full templates.
+
+Implementations MAY support parsing partial templates and matching portions of the paths against partial templates, as well as combining the templates.
+
+==Parsing result==
+
+The result of successful parsing of a valid path template can be represented by a list of sections, where each section is a list of index ranges, where index range is a tuple of (range_start, range_end).
+
+==Matching==
+
+The matching is to be performed against a list of integer values that represent a BIP32 path (or a portion of BIP32 path, for partial templates).
+
+Non-hardened indexes in this list should be represented by values from 0 to 2147483647.
+
+Hardened indexes in this list should be represented by values from 2147483648 to 4294967295.
+
+The matching algorithm:
+
+    1. If the length of the path differs from the length of the template, fail
```

Comment:
Could you explain this for me? You used "path template" previously in the BIP (for example #1 and #8 under constraints), so I am confused by the different definitions of path and template?

Reply:
Do you mean section? as defined in the 3rd line under #Specification?

Reply:
"path" is something that is being matched, like `44'/0'/0'/1/5`, the "path template", or simply "template" in this case, is something that the path is matched against, like `44'/0'/0'/{0-1}/{0-5000}`. Since the length of the template (number of sections in it) is the same as the length of the path (number of indexes in it), this matching step passes. If the template would be just  `44'/0'/0'`, for example, the matching would fail immediately. The implementation could implement partial match, of course, by taking a portion of the path of the same length as the template, and matching it, but I don't think this needs to be specified.

Reply:
Length has a lot of definitions, and was confusing.  You have already defined sections. So maybe re-word as:
If the number of sections of the path differs from the number of sections of the path template, fail.
Also might want to include that definition of path in the BIP

Reply:
Added the definitions of "length of the template", "length of the path": https://github.com/dgpv/bips/commit/abc113edd4bc2339c6b64689c0b33635615c297e

I don't think that we should included the definition of the path. Throughout the text of the BIP, "BIP32 derivation path" is mentioned many times, and it is obvious that it is the main thing that the BIP itself is focused on.

BIP32 or BIP43 do not define "path", while BIP44 just goes and uses the term "path" straight away without additional definition. I think we can just leave it as it is and not add specific definition for the path.

But if you have a good idea for such definition, it may not hurt to add it.